### DISCUSSION TITLE
fetchUserInfoWithToken() return null after successful Google login

### DISCUSSION THREAD

**Discussion by AlbertoMoricca** [datamweb/shield-oauth#25] [Q&A] [answered]
Hi,

I have the following error after a successful Google login:

Datamweb\ShieldOAuth\Libraries\GoogleOAuth::fetchUserInfoWithToken(): Return value must be of type object, null returned

I check the code and I see that it call a Google Api with the token obtained after the login and sent to the call-back ShieldOAuth GET method.

Do you have any suggestion in order to understand the issue?

Furthermore, one of the request header is 'User-Agent', that has the value 'ShieldOAuth/1.0'. I thought that the issue may be that the application name on Google configuration didn't match this string, but I renamed my app in the  OAuth consent screen section and the result was the same.

Thanks,
Alberto

**Comment by datamweb** [comment]
Hello, are you using the latest version?
If not, please upgrade:
```console
composer require datamweb/shield-oauth:dev-develop
```

**Reply by AlbertoMoricca** [reply]
Hi,

My version of the library is updated.

Following the screen:

![immagine](https://user-images.githubusercontent.com/105434937/206997748-d9d11f1d-5a21-4c66-bbf0-8e0bb0de7ae6.png)

**Comment by AlbertoMoricca** [comment]
Hi,

Following the "[Get the keys](https://github.com/datamweb/shield-oauth/blob/develop/docs/get_keys.md)" guide, I'm in doubt about the needed application scopes in order to perform a successful login.

The scopes that I configured are:
- .../auth/userinfo.email
- .../auth/userinfo.profile
- openid

At any cases, I tried this scopes and no scopes configured. The result is the same.

**Reply by datamweb** [reply]
@AlbertoMoricca , I will check **after 2022-12-19** this issue carefully and tag you.

**Reply by AlbertoMoricca** [reply]
Ok, thanks a lot. Furthermore, I Will try to do another checks in order to provide you more info.

**Comment by AlbertoMoricca** [comment]
Hi,

I have another info for you.

I tried the Github login, and, after a successful login, i had the following error:

CodeIgniter\Shield\Entities\User::setEmail(): Argument #1 ($email) must be of type string, null given, called in /workspace/vendor/codeigniter4/framework/system/Entity/Entity.php on line 451

It seems that the issue is not only on Google login and maybe the issues are related: both happens after a successful login.

Thanks in advice.

**Comment by AlbertoMoricca** [comment]
Hi,

I analyzed the code of GoogleOAuth. I tried manually (with Postman) the requests performed by the code and they works all fine.

**The same request with the same parameter** (as it seems) generate the error that i wrote in the first post:

"Datamweb\ShieldOAuth\Libraries\GoogleOAuth::fetchUserInfoWithToken(): Return value must be of type object, null returned"

The real reason of this null is a bad request returned by the server, without a more specific details.

Hoping that these details help.

Alberto

**Reply by datamweb** [reply]
Hi, thanks for being here and sharing your observations.

At the same time as the report, at first glance I did not notice the problem. So I took a break from code for a while.

Unfortunately, I have had a few surgeries recently, and will be having one almost every two weeks from now on. For this reason, I cannot concentrate and work with the PC because of the pain.

As soon as my physical condition improves, I will definitely pursue this issue with high concentration.
Sorry again for the delay.

**Reply by AlbertoMoricca** [reply]
Ok, thanks for your answer and don't worry for your issue. The health is more important of all. Have a good recovery.
If I will discover other details, I will share them.

**Reply by ambikaaSharma** [reply]
We resolve this error by changing Curl Request form "GET" to "POST",

**Comment by ambikaaSharma** [comment, marked as answer]
Hi,

**Datamweb\ShieldOAuth\Libraries\GoogleOAuth::fetchUserInfoWithToken(): Return value must be of type object, null returned**


We resolve this error by changing Curl Request form "GET" to "POST", 


`PROJECT-PATH/vendor/datamweb/shield-oauth/src/Libraries/GoogleOAuth.php`


```php
 protected function fetchUserInfoWithToken(): object
    {
        // send request to API URL
        try {
            $response = $this->client->request('**POST**', self::$API_USER_INFO_URL, [
                'headers' => [
                    'Accept'        => 'application/json',
                    'User-Agent'    => self::$APPLICATION_NAME . '/1.0',
                    'Authorization' => 'Bearer ' . $this->getToken(),
                ],
                'http_errors' => false,
            ]);
        } catch (Exception $e) {
            exit($e->getMessage());
        }

        return json_decode($response->getBody());
    }
```

**Reply by datamweb** [reply]
@ambikaaSharma Thank you for your contribution to this topic.

**Reply by AlbertoMoricca** [reply]
@ambikaaSharma Thanks for your solution. I tried with a custom copy of the Google Authenticator (because I can't edit the library installed by composer) and now the error is "Typed property CodeIgniter\Shield\Models\BaseModel::$tables must not be accessed before initialization". I think that the original issue is resolved, then.

**Reply by ambikaaSharma** [reply]
Hi

Please add  initialize in ShieldOAuthModel
(/var/www/html/projectfolder/vendor/datamweb/shield-oauth/src/Models/ShieldOAuthModel.php) on line number 22


> parent::initialize();


Thanks

**Reply by AlbertoMoricca** [reply]
Hi,

your solution resolve the issue.

Thanks a lot.

**Comment by AlbertoMoricca** [comment]
I tried to apply the fixes suggested by ambikaaSharma on a new branch but I have not the write permissions for a new branch or for push the changes on develop. Then, What's the procedure for ask to appy the bugfixes on the library?

The changes needed are:
1) /var/www/html/projectfolder/vendor/datamweb/shield-oauth/src/Models/ShieldOAuthModel.php) add on line number 22:
parent::initialize();

2) Change GET to POST in fetchUserInfoWithToken method in file PROJECT-PATH/vendor/datamweb/shield-oauth/src/Libraries/GoogleOAuth.php

Thanks in advice

**Reply by datamweb** [reply]
> 1. /var/www/html/projectfolder/vendor/datamweb/shield-oauth/src/Models/ShieldOAuthModel.php) add on line number 22:
>    parent::initialize();

@AlbertoMoricca The cause of problem **1** is related to a new feature that was recently added to Shield, so the suggested solution should definitely be applied to **shield-oauth**.

> 2\. Change GET to POST in fetchUserInfoWithToken method in file PROJECT-PATH/vendor/datamweb/shield-oauth/src/Libraries/GoogleOAuth.php

Regarding this, I'm not sure where the problem started, so I haven't taken any action.

As a result, you can change the first option for now. To participate in the project, please see [here](https://docs.github.com/en/get-started/quickstart/contributing-to-projects). 
Also note that **ALL** commits must be signed, see [here](https://github.com/datamweb/shield-oauth/pull/17#issuecomment-1308549694) for more information.

Let me know if you need help.
